crml/sylo/src/device.rs

```diff
@@ -13,49 +13,108 @@
-		DeviceAdded(AccountId, Hash, u32),
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// There are no devices registered for user (missing user_id in Devices)
+		UserId,
+		/// Device is already registered to user (device_id is already in use)
+		DeviceId,
+		/// A user can't have more than MAX_DEVICES registered devices
+		MaxDevice,
 	}
-);
+}
 
 impl<T: Trait> Module<T> {
-	pub fn append_device(user_id: &T::AccountId, device_id: u32) -> DispatchResult {
+	pub fn append_device(user_id: &T::AccountId, device_id: DeviceId) -> DispatchResult {
 		let mut devices = <Devices<T>>::get(user_id);
 
-		ensure!(!devices.contains(&device_id), "Device Id already in use");
-		ensure!(
-			devices.len() <= MAX_DEVICES,
-			"User has registered up to the maximum number of devices"
-		);
+		ensure!(!devices.contains(&device_id), Error::<T>::DeviceId);
+		ensure!(devices.len() < MAX_DEVICES, Error::<T>::MaxDevice);
 
 		devices.push(device_id);
 
 		<Devices<T>>::insert(user_id, devices);
 
 		Ok(())
 	}
+
+	pub fn delete_device(user_id: &T::AccountId, device_id: DeviceId) -> DispatchResult {
+		ensure!(<Devices<T>>::contains_key(user_id), Error::<T>::UserId);
+		let mut devices = <Devices<T>>::take(user_id);
+		devices.retain(|device| *device != device_id);
+		<Devices<T>>::insert(user_id, devices);
+		Ok(())
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::mock::{ExtBuilder, Test};
+	use frame_support::{assert_noop, assert_ok};
+	use sp_core::H256;
+
+	type Device = Module<Test>;
+
+	#[test]
+	fn append_device_works() {
+		ExtBuilder.build().execute_with(|| {
+			let user_id = H256::from_low_u64_be(1);
+			let device_id = 7357;
+
+			// simple appending of a device works
+			assert_ok!(Device::append_device(&user_id, device_id));
+			assert_eq!(Device::devices(user_id), vec![device_id]);
+
+			// adding the same device should return error
+			assert_noop!(Device::append_device(&user_id, device_id), Error::<Test>::DeviceId);
+
+			// adding more than MAX_DEVICES is not allowed
+			let new_devices: Vec<_> = ((device_id + 1)..(device_id + MAX_DEVICES as DeviceId)).collect();
+			assert_eq!(new_devices.len(), 999); // length assert is here in case MAX_DEVICES changes
+			for new_device in new_devices.clone() {
+				assert_ok!(Device::append_device(&user_id, new_device));
+			}
+			assert_eq!(Device::devices(user_id).len(), 1000);
+			assert_noop!(Device::append_device(&user_id, 123), Error::<Test>::MaxDevice);
+		});
+	}
+
+	#[test]
+	fn delete_device_works() {
+		ExtBuilder.build().execute_with(|| {
+			let user_id = H256::from_low_u64_be(1);
+			let mut devices = vec![1, 2, 3, 4, 5];
+			for device in devices.clone() {
+				assert_ok!(Device::append_device(&user_id, device));
+			}
+			for device in devices.clone().into_iter().rev() {
+				assert_ok!(Device::delete_device(&user_id, device));
+				devices.pop().unwrap();
+				assert_eq!(Device::devices(user_id), devices);
+			}
+			assert_eq!(Device::devices(user_id), vec![]);
+		});
```

Comment:
[optional] Could be nice to add a check that deleing a device that isn't on the list doesn't affect the list.
eg: `Device::devices(user_id) = [1,2,3,4]; Device::delete_device(&user_id, 5); assert_eq!(Device::devices(user_id), [1,2,3,4]);`

Reply:
Added `fn delete_non_existing_device_works` to show deleting a non-existing device passes through without an error